### PR TITLE
Sucesive backstack operation bug fix

### PR DESCRIPTION
**Pull request by SamuPS**

### CODE REVIEW

File: libraries/rib-base/src/androidTest/java/com/badoo/ribs/android/dialogkeyboard/AlertDialogKeyboardTest.kt [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+import com.badoo.ribs.test.util.ribs.root.TestRootRouter
+import com.badoo.ribs.test.util.ribs.root.builder.TestRootBuilder
+import com.badoo.ribs.test.util.runOnMainSync
+import com.badoo.ribs.test.util.waitFor
+import org.hamcrest.CoreMatchers.endsWith
+import org.junit.Rule
+import org.junit.Test
+import java.util.*
+
+class AlertDialogKeyboardTest {
```
Comment (SamuPS) [inline libraries/rib-base/src/androidTest/java/com/badoo/ribs/android/dialogkeyboard/AlertDialogKeyboardTest.kt:25]: This is already in master, this shouldn't be showed in the diff.

File: libraries/rib-base/src/androidTest/java/com/badoo/ribs/android/lifecycle/BaseNodesTest.kt [hunk 2]
```diff
@@ -43,8 +44,8 @@ abstract class BaseNodesTest {
         var backStack: BackStack<TestRootRouter.Configuration>? = null
 
         ribsRule.start { activity, savedInstanceState ->
-            val buildParams = BuildParams(
-                payload = null,
+            val buildParams = BuildParams<TestRootBuilder.Params>(
```
Comment (SamuPS) [inline libraries/rib-base/src/androidTest/java/com/badoo/ribs/android/lifecycle/BaseNodesTest.kt:47]: This is already in master, this shouldn't be showed in the diff.

File: libraries/rib-base/src/androidTest/java/com/badoo/ribs/android/lifecycle/RootNodeLifecycleTest.kt [hunk 3]
```diff
@@ -16,7 +19,13 @@ class RootNodeLifecycleTest {
     @get:Rule
     val ribsRule = RibsRule { activity, savedInstanceState ->
         provider.create(
-            buildParams = emptyBuildParams(),
+            buildParams = BuildParams(
+                payload = TestRootBuilder.Params(false),
```
Comment (SamuPS) [inline libraries/rib-base/src/androidTest/java/com/badoo/ribs/android/lifecycle/RootNodeLifecycleTest.kt:23]: This is already in master, this shouldn't be showed in the diff.

File: libraries/rib-base/src/androidTest/java/com/badoo/ribs/test/util/ribs/child/TestChildInteractor.kt [hunk 4]
```diff
@@ -4,7 +4,7 @@ import com.badoo.ribs.clienthelper.interactor.Interactor
 import com.badoo.ribs.core.modality.BuildParams
 
 class TestChildInteractor(
-    buildParams: BuildParams<Nothing?>
+    buildParams: BuildParams<*>
```
Comment (SamuPS) [inline libraries/rib-base/src/androidTest/java/com/badoo/ribs/test/util/ribs/child/TestChildInteractor.kt:7]: This is already in master, this shouldn't be showed in the diff.

File: libraries/rib-base/src/androidTest/java/com/badoo/ribs/test/util/ribs/child/TestChildView.kt [hunk 5]
```diff
@@ -3,18 +3,35 @@ package com.badoo.ribs.test.util.ribs.child
...
+import android.widget.EditText
 import android.widget.FrameLayout
 import com.badoo.ribs.core.view.AndroidRibView
 import com.badoo.ribs.core.view.RibView
 
 interface TestChildView : RibView
 
 class TestChildViewImpl(
-    context: Context, attrs: AttributeSet? = null, defStyle: Int = 0
+    context: Context,
```
Comment (SamuPS) [inline libraries/rib-base/src/androidTest/java/com/badoo/ribs/test/util/ribs/child/TestChildView.kt:14]: This is already in master, this shouldn't be showed in the diff.

File: libraries/rib-base/src/androidTest/java/com/badoo/ribs/test/util/ribs/child/TestChildView.kt [hunk 6]
```diff
@@ -3,18 +3,35 @@ package com.badoo.ribs.test.util.ribs.child
...
+    addEditText: Boolean,
+    attrs: AttributeSet? = null,
+    defStyle: Int = 0
 ) : AndroidRibView(),
     TestChildView {
 
     override val androidView: ViewGroup by lazy {
-        FrameLayout(context, attrs, defStyle)
+        FrameLayout(context, attrs, defStyle).also {
```
Comment (SamuPS) [inline libraries/rib-base/src/androidTest/java/com/badoo/ribs/test/util/ribs/child/TestChildView.kt:22]: This is already in master, this shouldn't be showed in the diff.

File: libraries/rib-base/src/androidTest/java/com/badoo/ribs/test/util/ribs/child/builder/TestChildBuilder.kt [hunk 7]
```diff
@@ -10,16 +10,23 @@ import com.badoo.ribs.test.util.ribs.child.TestChildRouter
 import com.badoo.ribs.test.util.ribs.child.TestChildView
 import com.badoo.ribs.test.util.ribs.child.TestChildViewImpl
 
-class TestChildBuilder : SimpleBuilder<TestNode<TestChildView>>() {
+class TestChildBuilder : Builder<TestChildBuilder.Params, TestNode<TestChildView>>() {
 
-    override fun build(buildParams: BuildParams<Nothing?>): TestNode<TestChildView> {
+    class Params(
```
Comment (SamuPS) [inline libraries/rib-base/src/androidTest/java/com/badoo/ribs/test/util/ribs/child/builder/TestChildBuilder.kt:15]: This is already in master, this shouldn't be showed in the diff.

File: libraries/rib-base/src/androidTest/java/com/badoo/ribs/test/util/ribs/root/TestRoot.kt [hunk 8]
```diff
@@ -45,27 +45,30 @@ interface TestRoot : Rib {
...
             private set
 
-        private fun builder(block: (TestNode<TestChildView>) -> Unit): (BuildContext) -> TestNode<TestChildView> =
-            {
-                TestChildBuilder().build(it).also {
-                    block.invoke(it)
-                }
+        private fun builder(
+            addEditText: Boolean,
+            block: (TestNode<TestChildView>) -> Unit
```
Comment (SamuPS) [inline libraries/rib-base/src/androidTest/java/com/badoo/ribs/test/util/ribs/root/TestRoot.kt:50]: This is already in master, this shouldn't be showed in the diff.

File: libraries/rib-base/src/androidTest/java/com/badoo/ribs/test/util/ribs/root/TestRootInteractor.kt [hunk 9]
```diff
@@ -9,7 +9,7 @@ import com.badoo.ribs.core.modality.BuildParams
 import io.reactivex.observers.TestObserver
 
 class TestRootInteractor(
-    buildParams: BuildParams<Nothing?>,
+    buildParams: BuildParams<*>,
```
Comment (SamuPS) [inline libraries/rib-base/src/androidTest/java/com/badoo/ribs/test/util/ribs/root/TestRootInteractor.kt:12]: This is already in master, this shouldn't be showed in the diff.

File: libraries/rib-base/src/androidTest/java/com/badoo/ribs/test/util/ribs/root/builder/TestRootBuilder.kt [hunk 10]
```diff
@@ -12,9 +12,13 @@ import com.badoo.ribs.test.util.ribs.root.TestRootViewImpl
 
 class TestRootBuilder(
     private val dependency: TestRoot.Dependency
-) : SimpleBuilder<Node<TestRootView>>() {
+) : Builder<TestRootBuilder.Params, Node<TestRootView>>() {
 
-    override fun build(buildParams: BuildParams<Nothing?>): Node<TestRootView> {
+    class Params(
+        val addEditText: Boolean = false
+    )
```
Comment (SamuPS) [inline libraries/rib-base/src/androidTest/java/com/badoo/ribs/test/util/ribs/root/builder/TestRootBuilder.kt:19]: This is already in master, this shouldn't be showed in the diff.

File: libraries/rib-base/src/main/java/com/badoo/ribs/android/dialog/AlertDialogLauncher.kt [hunk 11]
```diff
@@ -16,6 +16,7 @@ class AlertDialogLauncher(
     init {
         lifecycle.subscribe(onDestroy = {
             dialogs.values.forEach { it.dismiss() }
+            dialogs.clear()
```
Comment (SamuPS) [inline libraries/rib-base/src/main/java/com/badoo/ribs/android/dialog/AlertDialogLauncher.kt:19]: This is already in master, this shouldn't be showed in the diff.

File: libraries/rib-base/src/main/java/com/badoo/ribs/android/dialog/AlertDialogLauncher.kt [hunk 12]
```diff
@@ -27,6 +28,7 @@ class AlertDialogLauncher(
 
     override fun hide(dialog: Dialog<*>) {
         dialogs[dialog]?.dismiss()
+        dialogs.remove(dialog)
```
Comment (SamuPS) [inline libraries/rib-base/src/main/java/com/badoo/ribs/android/dialog/AlertDialogLauncher.kt:31]: This is already in master, this shouldn't be showed in the diff.

File: libraries/rib-base/src/main/java/com/badoo/ribs/android/dialog/DialogExtensions.kt [hunk 13]
```diff
@@ -22,6 +22,7 @@ fun <Event : Any> Dialog<Event>.toAlertDialog(context: Context, onClose: () -> U
             .apply {
                 setCanceledOnTouchOutside(this@toAlertDialog)
                 setButtonClickListeners(this@toAlertDialog, onClose)
+                setOnDismissListener { this@toAlertDialog.rib = null }
```
Comment (SamuPS) [inline libraries/rib-base/src/main/java/com/badoo/ribs/android/dialog/DialogExtensions.kt:25]: This is already in master, this shouldn't be showed in the diff.

File: libraries/rib-base/src/main/java/com/badoo/ribs/routing/state/feature/Actor.kt [hunk 14]
```diff
@@ -218,5 +212,29 @@ internal class Actor<C : Parcelable>(
...
+        when {
+            pendingTransition.isScheduled(state) && canTransition(params) -> pendingTransition.execute(requireNotNull(transitionHandler)).start()
+            pendingTransition.isScheduled(state) && canTransition(params).not() -> pendingTransition.completeWithoutTransition()
+            else -> pendingTransition.discard()
+        }
+    }
+
+    private fun <C : Parcelable> PendingTransition<C>.isScheduled(state: WorkingState<C>) = this in state.pendingTransitions
+
+    private fun canTransition(params: TransactionExecutionParams<C>) = params.globalActivationLevel != SLEEPING && transitionHandler != null
```
Comment (zsoltk) [inline libraries/rib-base/src/main/java/com/badoo/ribs/routing/state/feature/Actor.kt:239]:
```suggestion
    private fun <C : Parcelable> PendingTransition<C>.isScheduled(state: WorkingState<C>) = 
        this in state.pendingTransitions

    private fun canTransition(params: TransactionExecutionParams<C>) = 
        params.globalActivationLevel != SLEEPING && transitionHandler != null
```

File: libraries/rib-base/src/main/java/com/badoo/ribs/routing/state/feature/PendingTransitionFactory.kt [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+import com.badoo.ribs.routing.state.changeset.TransitionDescriptor
+import com.badoo.ribs.routing.transition.TransitionDirection
+import com.badoo.ribs.routing.transition.TransitionElement
+
+internal class PendingTransitionFactory<C : Parcelable>(
+    private val effectEmitter: EffectEmitter<C>,
+    private val internalTransactionConsumer: InternalTransactionConsumer<C>
+) {
+
+    fun make(
```
Comment (zsoltk) [inline libraries/rib-base/src/main/java/com/badoo/ribs/routing/state/feature/PendingTransitionFactory.kt:14]: `create`?

File: libraries/rib-base/src/main/java/com/badoo/ribs/routing/state/feature/RoutingStatePool.kt [hunk 16]
```diff
@@ -195,8 +215,19 @@ internal class RoutingStatePool<C : Parcelable>(
...
+        discardPendingTransitions(state.pendingTransitions)
+    }
+
+    private fun disposeOngoingTransitions(ongoingTransitions: List<OngoingTransition<C>>) {
+        ongoingTransitions.forEach {
             it.dispose()
         }
     }
+
+    private fun discardPendingTransitions(pendingTransition: List<PendingTransition<C>>) {
```
Comment (zsoltk) [inline libraries/rib-base/src/main/java/com/badoo/ribs/routing/state/feature/RoutingStatePool.kt:228]:
```suggestion
    private fun cancelPendingTransitions(pendingTransition: List<PendingTransition<C>>) {
```

File: libraries/rib-base/src/test/java/com/badoo/ribs/routing/state/feature/ActorTest.kt [hunk 17]
```diff
@@ -0,0 +1,393 @@
...
+            descriptor = mock()
+        )
+
+        getActor().invoke(state, transaction)
+
+        verify(pendingTransition).schedule()
+    }
+
+    /**
+     * [Case-1]
```
Comment (zsoltk) [inline libraries/rib-base/src/test/java/com/badoo/ribs/routing/state/feature/ActorTest.kt:39]:
```suggestion
     * [Case-2]
```

**Review by zsoltk — CHANGES_REQUESTED** [review]

**Review by zsoltk — APPROVED** [review]

**Pull request merged**